File: test/test_simulate.go
package main

import (
	"fmt"
	"time"

	"github.com/gin-gonic/gin"
	"github.com/jackc/pgx"
)

// Database connectivity variables
var db *pgx.ConnPool
var db_err error

//Initialise connection to the database
func init() {
	db, db_err = pgx.NewConnPool(pgx.ConnPoolConfig{
		ConnConfig: pgx.ConnConfig{
			Host:     "localhost",
			Database: "pmp",
			User:     "ashutosh",
			Password: "123",
			Port:     5432,
		},
		MaxConnections: 10,
	})

	if db_err != nil {
		fmt.Println("Can't connect to database")
	}
}

func main() {
	r := gin.Default()
	r.POST("/send", func(c *gin.Context) {
		var request Request
		// Dump the incoming request data into a struct variable
		c.BindJSON(&request)
		fmt.Println("\n\n\n Received : %#v\n\n\n", request)
		// database transaction begins

		tx, _ := db.Begin() // tx => transaction , _ => error and execute

		defer tx.Rollback() // it will be executed after the completion of local function

		var userid int64 //variable for storing user_id
		var deviceid string
		var platform string
		// transaction query for inserting user_name and name and it will return user_id

		// tx.QueryRow(`
		//     INSERT INTO alpha (user_name, name) values ($1, $2) returning user_id
		// `, request.Username, request.Name).Scan(&user_id)

		// Fetch the user_id for the corresponding input request
		fmt.Println("Username and email Randomly generated from test_simulate.py : ", request.Username, request.Email)
		// db.QueryRow(`
		// 		SELECT userid
		// 		FROM users
		// 		WHERE username = $1 and email = $2`, request.Username, request.Email).Scan(&userid)
		// fmt.Print("User id after query : ", userid)

		// // Select the device corresponding to the above user_id
		// err := db.QueryRow(`
		// 		SELECT deviceid, platform
		// 		FROM usersdescription
		// 		WHERE usersdescription.userid = $1`, userid).Scan(&deviceid, &platform)

		// natural join of above two commands
		err := db.QueryRow(`
                SELECT usersdescription.deviceid, usersdescription.platform
                FROM users, usersdescription
                WHERE usersdescription.userid = users.userid AND username = $1 AND email = $2
        `, request.Username, request.Email).Scan(&deviceid, &platform)

		fmt.Println("\n\nDeviceid and platform : ", deviceid, platform)
		fmt.Println(err)

		// commit_err := tx.Commit() // commit the transaction

		// if commit_err != nil {
		// 	tx.Rollback()
		// 	c.JSON(500, "Duplicate insertion") // commit error should be mentioned according to the corresponding errors
		// 	return
		// }

		var response Response
		response.UserID = userid
		response.DeviceID = deviceid
		response.Timestamp = time.Now().Unix()
		response.Platform = platform
		fmt.Println("\n\n\n Response :  %#v\n\n\n", response)
		c.JSON(200, response)
	})

	// r.POST("/gogs", func(c *gin.Context) {
	// 	m := make(map[string]interface{})
	// 	c.BindJSON(&m)
	// 	fmt.Printf("\n\n\nReceived new push : %#v\n\n\n\n", m)
	// 	c.JSON(200, "ok")
	// })

	fmt.Print("Going live on :7070\n")
	r.Run(":7070")
}

// Request holds
type Request struct {
	Username string `json:"username,omitempty"` // username
	Name     string `json:"name,omitempty"`
	Email    string `json:"email,omitempty"`
}

//Response
type Response struct {
	UserID    int64  `json:"userid,omitempty"`
	DeviceID  string `json:"deviceid,omitempty"`
	Timestamp int64  `json:"timestamp,omitempty"`
	Platform  string `json:"platform,omitempty"`
}

File: test/test_simulate.py
import psycopg2
import random
import requests
import json

postgres = psycopg2.connect(database = 'pmp', host = 'localhost', user = 'ashutosh', password = '123')
cursor = postgres.cursor()

#   generate simulation module for users{relation} here user_id is unique and 
#   it will be used for the purpose of mapping from user_description{relation} and
#   this relation will have the information of user details like which device is 
#   used recently and corresponding to that it will send the push notifications
#   In user_description relation we will create a primary key user_id and on the 
#   basis of this very primary key we will map to the users and after mapping we
#   will send the push notification to the concerned device with the appropriate 
#   server i.e. {use apns for apple, gcm for android etc}.
 
 
 ######################################################################################
 ######################################################################################
 #                               CHECK FOR TABLE USERS                                #
 ######################################################################################
 ###################################################################################### 
 
cursor.execute("""
   select exists(select 0 from pg_class where relname = 'users')
""") 
# it returns true if there is at least row in database and it will be a tuple
presence = cursor.fetchone()[0]
print(presence)



 ######################################################################################
 ######################################################################################
 #                               DELETE TABLE USERS IF EXISTS                         #
 ######################################################################################
 ######################################################################################
 
if presence:
    print('users table exists, deleting here')
    cursor.execute("""
        drop table users
    """)



 ######################################################################################
 ######################################################################################
 #                               CREATE TABLE USERS                                   #
 ######################################################################################
 ######################################################################################
cursor.execute("""
    create table users( 
        userid bigserial,
        username text,
        name text,
        email text
    )
""")
a = ["abs", "xyz", "vrt", "alpha", "beta"]
b = {"abs":"as01hu@gmail.com","xyz":"xyz@outlook.com","vrt": "alpha@yahoomail.com", "alpha":"xwe@yahoo.in", "beta":"ade@gmail.com"}

for i in range(5):
    username = random.choice(a)
    name = random.choice(["Ashu", "pp"])
    email = b[username]
    a.remove(username)
    cursor.execute("""
        INSERT INTO users (username, name, email) values(%s, %s, %s)
    """,(username, name, email))

######################################################################################
######################################################################################
#                       COMMIT AND ROLLBACK IF EXCEPTION                             #
######################################################################################
######################################################################################

try:
    postgres.commit()
except Exception as e:
    postgres.rollback()
    print(e)


 ######################################################################################
 ######################################################################################
 #                               CHECK FOR TABLE USERSDESCRIPTION                     #
 ######################################################################################
 ######################################################################################
 
cursor.execute("""
   select exists(select 0 from pg_class where relname = 'usersdescription')
""") 
# it returns true if there is at least row in database and it will be a tuple
presence = cursor.fetchone()[0]
print(presence)


 ######################################################################################
 ######################################################################################
 #                       DELETE USERSDESCRIPTION IF EXISTS                            #
 ######################################################################################
 ######################################################################################
 
 
if presence:
    print('usersdescription table exists, deleting here')
    cursor.execute("""
        drop table usersdescription
    """)

 ######################################################################################
 ######################################################################################
 #                       CREATE TABLE USERSDESCRIPTION                                #
 ######################################################################################
 ######################################################################################

cursor.execute("""
    create table usersdescription( 
            userid bigint,
            deviceid text,
            platform text
            -- CONSTRAINT userdescription_pk PRIMARY KEY (userid)
    )
""")

######################################################################################
######################################################################################
#                               DUMP RANDOM DATA                                     #
######################################################################################
######################################################################################

for i in range(15):
    userid = i+1
    deviceid = random.choice(["gdjg#5r67fgvehjdg@7781uigd", "iljilwefhl%^%^ghfgwekuqeg2", "iouyoh6587123468ykjqhdkqdg"])
    platform = random.choice(["ios", "apple", "windows"]) 
    
    cursor.execute("""
        INSERT INTO usersdescription (userid, deviceid, platform) values(%s, %s, %s)
    """,(userid, deviceid, platform))
    
##########          check what is generated          ###############################
    
#####################################################################################
#####################################################################################    
# cursor.execute("""   
#    select * from usersdescription 
# """)
#####################################################################################
#####################################################################################


######################################################################################
######################################################################################
#                       COMMIT AND ROLLBACK IF EXCEPTION                             #
######################################################################################
######################################################################################

try:
    postgres.commit()
except Exception as e:
    postgres.rollback()
    print(e)
    
a = ["abs", "xyz", "vrt", "alpha", "beta"]
b = {"abs":"as01hu@gmail.com","xyz":"xyz@outlook.com","vrt": "alpha@yahoomail.com", "alpha":"xwe@yahoo.in", "beta":"ade@gmail.com"}

for i in range(5):
    username = random.choice(a)
    name = random.choice(["Ashu", "pp"])
    email = b[username]
    a.remove(username)
    message_send = {
        "username" : username,
        "name" : name,
        "email" : email
    }
    resp = requests.post("http://localhost:7070/send", data = json.dumps(message_send))
    body = resp.text
    print(body)
    
    
    # cursor.execute("""   
    #     INSERT INTO alpha (user_name, name) values (%s, %s)
    # """, (user_name, name))
	
# try:
#     postgres.commit()
# except Exception as e:
#     postgres.rollback()
#     print(e)

# cursor.execute("""   
#    select * from alpha 
# """)

# a = cursor.fetchall()
# for row in a : 
#     print(row)
